nifi-nar-bundles/nifi-cassandra-bundle/nifi-cassandra-processors/src/main/java/org/apache/nifi/processors/cassandra/PutCassandraQL.java

```diff
@@ -0,0 +1,383 @@
+        return descriptors;
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+
+    @OnScheduled
+    public void onScheduled(final ProcessContext context) {
+        ProcessorLog log = getLogger();
+        try {
+            connectToCassandra(context);
+        } catch (final NoHostAvailableException nhae) {
+            log.error("No host in the Cassandra cluster can be contacted successfully to execute this statement", nhae);
+            // Log up to 10 error messages. Otherwise if a 1000-node cluster was specified but there was no connectivity,
+            // a thousand error messages would be logged. However we would like information from Cassandra itself, so
+            // cap the error limit at 10, format the messages, and don't include the stack trace (it is displayed by the
+            // logger message above).
+            log.error(nhae.getCustomMessage(10, true, false));
+            throw new ProcessException(nhae);
+        } catch (final AuthenticationException ae) {
+            log.error("Invalid username/password combination", ae);
+            throw new ProcessException(ae);
+        }
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        ProcessorLog logger = getLogger();
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        final long startNanos = System.nanoTime();
+        final long statementTimeout = context.getProperty(STATEMENT_TIMEOUT).asTimePeriod(TimeUnit.MILLISECONDS);
+        final Charset charset = Charset.forName(context.getProperty(CHARSET).getValue());
+
+        // The documentation for the driver recommends the session remain open the entire time the processor is running
+        // and states that it is thread-safe. This is why connectionSession is not in a try-with-resources.
+        final Session connectionSession = cassandraSession.get();
+
+        String cql = getCQL(session, flowFile, charset);
+        try {
+            PreparedStatement statement = connectionSession.prepare(cql);
+            BoundStatement boundStatement = statement.bind();
+
+            Map<String, String> attributes = flowFile.getAttributes();
+            for (final Map.Entry<String, String> entry : attributes.entrySet()) {
+                final String key = entry.getKey();
+                final Matcher matcher = CQL_TYPE_ATTRIBUTE_PATTERN.matcher(key);
+                if (matcher.matches()) {
+                    final int parameterIndex = Integer.parseInt(matcher.group(1));
+                    String paramType = entry.getValue();
+                    if (StringUtils.isEmpty(paramType)) {
+                        throw new ProcessException("Value of the " + key + " attribute is null or empty, it must contain a valid value");
```

Comment:
The ProcessException(s) are caught below and the flow file is transferred to failure.